compiler/z/codegen/BinaryEvaluator.cpp

```diff
@@ -1848,10 +1888,29 @@ OMR::Z::TreeEvaluator::tryToReplaceLongAndWithRotateInstruction(TR::Node * node,
             //    half and bottom half of the register), then it's better to use RISBG.
             //    This is because there are no NI** instructions allowing us to specify bits in
             //    the top half and bottom half of the register to zero out.
-
+               
             if (firstChild->getReferenceCount() > 1 || lZeros > 31 || tZeros > 31
                 || (lZeros > 0 && tZeros > 0))
                {
+                  // If the shift amount is NOT zero, then we must account for the shifted range
+                  // [1] If the resultant shifted msb in a right shift is shifted "too far" to the right, then we are basically zeroing the register. This optimization will not be used.
+                  // [2] Same case if the resultant lsb of a left shift is shifted out of range.
+                  // Eg[1]: 000...0011 >> 3. msb = 62, lsb = 63. If we right shift, then msb and lsb will both be > 63 (out of range)
+                  // Eg[2]: 11100...00 << 10. msb = 0, lsb = 2. If we left shift, then msb and lsb will both be < 0 (out of range)
+               if (shiftAmount < 0)
+                  {
+                  if (msBit - shiftAmount > 63)
+                     {
+                     return NULL;
+                     }
+                  }
+               else if (shiftAmount > 0)
+                  {
+                  if (lsBit - shiftAmount < 0)
+                     {
+                     return NULL;
+                     }
+                  }
```

Comment:
How about:

```
if ((shiftAmount < 0 && msBit - shiftAmount > 63) ||
    (shiftAmount > 0 && lsBit - shiftAmount < 0))
   {
   return NULL;
   }
```

instead of these 14 lines. Also the comment needs to be formatted. There seems to be an extra 3 spaces.